File: totaleconomy-common/src/test/java/com/ericgrandt/totaleconomy/common/data/BalanceDataTest.java
package com.ericgrandt.totaleconomy.common.data;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.ericgrandt.totaleconomy.common.TestUtils;
import com.ericgrandt.totaleconomy.common.data.dto.BalanceDto;
import com.zaxxer.hikari.HikariDataSource;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.UUID;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
public class BalanceDataTest {
    @Test
    @Tag("Unit")
    public void getBalance_WithBalanceFound_ShouldReturnBalance() throws SQLException {
        // Arrange
        Database databaseMock = mock(Database.class);
        Connection connectionMock = mock(Connection.class);
        PreparedStatement preparedStatementMock = mock(PreparedStatement.class);
        ResultSet resultSetMock = mock(ResultSet.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(connectionMock);
        when(connectionMock.prepareStatement(anyString())).thenReturn(preparedStatementMock);
        when(preparedStatementMock.executeQuery()).thenReturn(resultSetMock);
        when(resultSetMock.next()).thenReturn(true);
        when(resultSetMock.getBigDecimal("balance")).thenReturn(BigDecimal.TEN);

        BalanceData sut = new BalanceData(databaseMock);

        // Act
        BigDecimal actual = sut.getBalance(UUID.randomUUID(), 1);
        BigDecimal expected = BigDecimal.TEN;

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void getBalance_WithNoBalanceFound_ShouldReturnNull() throws SQLException {
        // Arrange
        Database databaseMock = mock(Database.class);
        Connection connectionMock = mock(Connection.class);
        PreparedStatement preparedStatementMock = mock(PreparedStatement.class);
        ResultSet resultSetMock = mock(ResultSet.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(connectionMock);
        when(connectionMock.prepareStatement(anyString())).thenReturn(preparedStatementMock);
        when(preparedStatementMock.executeQuery()).thenReturn(resultSetMock);
        when(resultSetMock.next()).thenReturn(false);

        BalanceData sut = new BalanceData(databaseMock);

        // Act
        BigDecimal actual = sut.getBalance(UUID.randomUUID(), 1);

        // Assert
        assertNull(actual);
    }

    @Test
    @Tag("Unit")
    public void updateBalance_WithBalanceUpdated_ShouldReturnOne() throws SQLException {
        // Arrange
        Database databaseMock = mock(Database.class);
        Connection connectionMock = mock(Connection.class);
        PreparedStatement preparedStatementMock = mock(PreparedStatement.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(connectionMock);
        when(connectionMock.prepareStatement(anyString())).thenReturn(preparedStatementMock);
        when(preparedStatementMock.executeUpdate()).thenReturn(1);

        BalanceData sut = new BalanceData(databaseMock);

        // Act
        int actual = sut.updateBalance(UUID.randomUUID(), 1, 100);
        int expected = 1;

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void updateBalance_WithNoBalanceUpdated_ShouldReturnZero() throws SQLException {
        // Arrange
        Database databaseMock = mock(Database.class);
        Connection connectionMock = mock(Connection.class);
        PreparedStatement preparedStatementMock = mock(PreparedStatement.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(connectionMock);
        when(connectionMock.prepareStatement(anyString())).thenReturn(preparedStatementMock);
        when(preparedStatementMock.executeUpdate()).thenReturn(0);

        BalanceData sut = new BalanceData(databaseMock);

        // Act
        int actual = sut.updateBalance(UUID.randomUUID(), 1, 100);
        int expected = 0;

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Integration")
    public void getBalance_ShouldReturnBalance() throws SQLException {
        // Arrange
        TestUtils.resetDb();
        TestUtils.seedCurrencies();
        TestUtils.seedAccounts();

        UUID accountId = UUID.fromString("62694fb0-07cc-4396-8d63-4f70646d75f0");
        int currencyId = 1;

        Database databaseMock = mock(Database.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(TestUtils.getConnection());

        BalanceData sut = new BalanceData(databaseMock);

        // Act
        BigDecimal actual = sut.getBalance(accountId, currencyId);
        BigDecimal expected = BigDecimal.valueOf(50).setScale(2, RoundingMode.DOWN);

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Integration")
    public void updateBalance_ShouldUpdateBalance() throws SQLException {
        // Arrange
        TestUtils.resetDb();
        TestUtils.seedCurrencies();
        TestUtils.seedAccounts();

        UUID accountId = UUID.fromString("62694fb0-07cc-4396-8d63-4f70646d75f0");
        int currencyId = 1;

        Database databaseMock = mock(Database.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(TestUtils.getConnection());

        BalanceData sut = new BalanceData(databaseMock);

        // Act
        int actual = sut.updateBalance(accountId, currencyId, 100);
        int expected = 1;

        BalanceDto actualBalanceDto = TestUtils.getBalanceForAccountId(accountId, currencyId);

        // Assert
        assertEquals(expected, actual);
        assertNotNull(actualBalanceDto);
        assertEquals(BigDecimal.valueOf(100).setScale(2, RoundingMode.DOWN), actualBalanceDto.balance());
    }

    @Test
    @Tag("Integration")
    public void transfer_ShouldUpdateBalances() throws SQLException {
        // Arrange
        TestUtils.resetDb();
        TestUtils.seedCurrencies();
        TestUtils.seedAccounts();

        Database databaseMock = mock(Database.class);
        when(databaseMock.getDataSource()).thenReturn(mock(HikariDataSource.class));
        when(databaseMock.getDataSource().getConnection()).thenReturn(TestUtils.getConnection());

        BalanceData sut = new BalanceData(databaseMock);

        UUID fromAccountId = UUID.fromString("62694fb0-07cc-4396-8d63-4f70646d75f0");
        UUID toAccountId = UUID.fromString("551fe9be-f77f-4bcb-81db-548db6e77aea");

        // Act
        sut.transfer(fromAccountId, toAccountId, 1, 10);

        BalanceDto actualFromBalanceDto = TestUtils.getBalanceForAccountId(fromAccountId, 1);
        BalanceDto actualToBalanceDto = TestUtils.getBalanceForAccountId(toAccountId, 1);
        BigDecimal expectedFromBalance = BigDecimal.valueOf(40).setScale(2, RoundingMode.DOWN);
        BigDecimal expectedToBalance = BigDecimal.valueOf(110).setScale(2, RoundingMode.DOWN);

        // Assert
        assertEquals(expectedFromBalance, actualFromBalanceDto.balance());
        assertEquals(expectedToBalance, actualToBalanceDto.balance());
    }
}

File: totaleconomy-common/src/test/kotlin/com/ericgrandt/totaleconomy/data/BalanceDataTest.kt
package com.ericgrandt.totaleconomy.data

import com.ericgrandt.totaleconomy.model.DepositIntoBalance
import com.ericgrandt.totaleconomy.model.SetBalance
import com.ericgrandt.totaleconomy.model.WithdrawFromBalance
import com.ericgrandt.totaleconomy.testutils.TestUtils
import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.getError
import io.mockk.MockKAnnotations
import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import org.junit.jupiter.api.Tag
import java.sql.SQLException
import java.util.UUID
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

class BalanceDataTest {
    @BeforeTest
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    @Tag("Integration")
    fun getBalance_WithBalance_ShouldReturnBalance() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()
        val testBalance = TestUtils.seedBalance(testAccount.id, null)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.getBalance(testAccount.id)
            val expected = Ok(testBalance)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun getBalance_WithNoBalance_ShouldReturnNull() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.getBalance(testAccount.id)
            val expected = Ok(null)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun getBalance_WithSqlException_ShouldReturnErrResult() {
        // Arrange
        TestUtils.connectToTestDb(false)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.getBalance(UUID.randomUUID())

            assertThat(actual.getError())
                .isInstanceOf(SQLException::class.java)
        }
    }

    @Test
    @Tag("Integration")
    fun setBalance_WithUpdatedRow_ShouldReturnOne() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()
        TestUtils.seedBalance(testAccount.id, null)

        val input = SetBalance(testAccount.id, 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.setBalance(input)
            val expected = Ok(1)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun setBalance_WithNoUpdatedRow_ShouldReturnZero() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()

        val input = SetBalance(testAccount.id, 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.setBalance(input)
            val expected = Ok(0)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun setBalance_WithSqlException_ShouldReturnErrResult() {
        // Arrange
        TestUtils.connectToTestDb(false)

        val input = SetBalance(UUID.randomUUID(), 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.setBalance(input)

            assertThat(actual.getError())
                .isInstanceOf(SQLException::class.java)
        }
    }

    @Test
    @Tag("Integration")
    fun withdrawFromBalance_WithUpdatedRow_ShouldReturnOne() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()
        TestUtils.seedBalance(testAccount.id, null)

        val input = WithdrawFromBalance(testAccount.id, 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.withdrawFromBalance(input)
            val expected = Ok(1)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun withdrawFromBalance_WithNoUpdatedRow_ShouldReturnZero() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()

        val input = WithdrawFromBalance(testAccount.id, 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.withdrawFromBalance(input)
            val expected = Ok(0)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun withdrawFromBalance_WithSqlException_ShouldReturnErrResult() {
        // Arrange
        TestUtils.connectToTestDb(false)

        val input = WithdrawFromBalance(UUID.randomUUID(), 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.withdrawFromBalance(input)

            assertThat(actual.getError())
                .isInstanceOf(SQLException::class.java)
        }
    }

    @Test
    @Tag("Integration")
    fun depositIntoBalance_WithUpdatedRow_ShouldReturnOne() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()
        TestUtils.seedBalance(testAccount.id, null)

        val input = DepositIntoBalance(testAccount.id, 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.depositIntoBalance(input)
            val expected = Ok(1)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun depositIntoBalance_WithNoUpdatedRow_ShouldReturnZero() {
        // Arrange
        TestUtils.connectToTestDb()
        val testAccount = TestUtils.seedAccount()

        val input = DepositIntoBalance(testAccount.id, 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.depositIntoBalance(input)
            val expected = Ok(0)

            assertEquals(expected, actual)
        }
    }

    @Test
    @Tag("Integration")
    fun depositIntoBalance_WithSqlException_ShouldReturnErrResult() {
        // Arrange
        TestUtils.connectToTestDb(false)

        val input = DepositIntoBalance(UUID.randomUUID(), 10.0)

        val sut = BalanceData()

        // Act/Assert
        transaction {
            val actual = sut.depositIntoBalance(input)

            assertThat(actual.getError())
                .isInstanceOf(SQLException::class.java)
        }
    }

    // @Test
    // @Tag("Integration")
    // fun transferBalance_WithSuccess_ShouldUpdateBalance() {
    //    // Arrange
    //    TestUtils.resetDb()
    //    val testFromAccount = TestUtils.seedAccount()
    //    TestUtils.seedBalance(testFromAccount.id, null)
    //    val testToAccount = TestUtils.seedAccount()
    //    TestUtils.seedBalance(testToAccount.id, null)

    //    every { databaseMock.dataSource } returns mockk<HikariDataSource>()
    //    every { databaseMock.dataSource.connection } answers { TestUtils.getConnection() }

    //    val input = TransferBalance(testFromAccount.id, testToAccount.id, 1.0)

    //    val sut = BalanceData(databaseMock)

    //    // Act
    //    val actual = sut.transferBalance(input)
    //    val expectedFromBalance = 0.23
    //    val expectedToBalance = 2.23

    //    // Assert
    //    val updatedFromBalance = sut.getBalance(testFromAccount.id).get()
    //    val updatedToBalance = sut.getBalance(testToAccount.id).get()

    //    assertEquals(Ok(true), actual)
    //    assertEquals(expectedFromBalance, updatedFromBalance?.balance)
    //    assertEquals(expectedToBalance, updatedToBalance?.balance)
    // }

    // @Test
    // @Tag("Integration")
    // fun transferBalance_WithSqlException_ShouldReturnErrResult() {
    //    // Arrange
    //    TestUtils.resetDb()
    //    val testFromAccount = TestUtils.seedAccount()
    //    val testToAccount = TestUtils.seedAccount()

    //    every { databaseMock.dataSource } returns mockk<HikariDataSource>()
    //    every { databaseMock.dataSource.connection } returns TestUtils.getConnection()
    //    every { databaseMock.dataSource.connection.prepareStatement(any()) } throws SQLException()

    //    val input = TransferBalance(testFromAccount.id, testToAccount.id, 10.0)

    //    val sut = BalanceData(databaseMock)

    //    // Act
    //    val actual = sut.transferBalance(input)

    //    // Assert
    //    assertThat(actual.getError())
    //        .isInstanceOf(SQLException::class.java)
    // }
}
